presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsAnnotationProcessor.java

```diff
@@ -0,0 +1,247 @@
+    }
+
+    private double processSingleArgumentStatistic(
+            List<Double> sourceStats,
+            int sourceStatsArgumentIndex,
+            StatsPropagationBehavior operation)
+    {
+        // sourceStatsArgumentIndex is index of the argument on which
+        // ScalarPropagateSourceStats annotation was applied.
+        double statValue = NaN;
+        if (operation.isMultiArgumentStat()) {
+            for (int i = 0; i < sourceStats.size(); i++) {
+                if (i == 0 && operation.isSourceStatsDependentStats() && isFinite(sourceStats.get(i))) {
+                    statValue = sourceStats.get(i); // initialize.
+                    continue;
+                }
+                switch (operation) {
+                    case MAX_TYPE_WIDTH_VARCHAR:
+                        statValue = maxExcludingNaNs(statValue, typeWidthVarchar(callExpression, i));
+                        break;
+                    case USE_MIN_ARGUMENT:
+                        statValue = min(statValue, sourceStats.get(i));
+                        break;
+                    case USE_MAX_ARGUMENT:
+                        statValue = max(statValue, sourceStats.get(i));
+                        break;
+                    case SUM_ARGUMENTS:
+                        statValue = statValue + sourceStats.get(i);
+                        break;
+                }
+            }
+        }
+        else {
+            switch (operation) {
+                case USE_SOURCE_STATS:
+                    statValue = sourceStats.get(sourceStatsArgumentIndex);
+                    break;
+                case ROW_COUNT:
+                    statValue = outputRowCount;
+                    break;
+                case NON_NULL_ROW_COUNT:
+                    statValue = outputRowCount * (1 - firstFiniteValue(nullFraction, 0.0));
+                    break;
+                case USE_TYPE_WIDTH_VARCHAR:
+                    statValue = typeWidthVarchar(callExpression, sourceStatsArgumentIndex);
+                    break;
+            }
+        }
+        return statValue;
+    }
+
+    private double typeWidthVarchar(CallExpression call, int argumentIndex)
+    {
+        TypeSignature typeSignature = call.getArguments().get(argumentIndex).getType().getTypeSignature();
+        if (typeSignature.getTypeSignatureBase().hasStandardType() && typeSignature.getTypeSignatureBase().getStandardTypeBase().equals(
+                StandardTypes.VARCHAR)) {
+            if (typeSignature.getParameters().size() == 1) { // Varchar type should have 1 parameter i.e. size.;
+                Long longLiteral = typeSignature.getParameters().get(0).getLongLiteral();
+                if (longLiteral > 0 && longLiteral != VarcharType.UNBOUNDED_LENGTH) {
+                    return longLiteral;
+                }
+            }
+        }
+        return NaN;
+    }
+
+    private double getReturnTypeWidth(CallExpression call, StatsPropagationBehavior operation)
```

Comment:
Again, it's surprising to see a double here. At first read, this really sounds like an integer to me. Why is this a double?

Reply:
I need it because, `VariableStatsEstimate` needs all the values as double. This way we avoid type conversions. https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/VariableStatsEstimate.java

Reply:
I think the reason all the members in `VariableStatsEstimate` (and related types like `PlanNodeStatsEstimate`) use doubles for integral properties is because this is the largest width primitive type we can use in Java (BigInteger could be used, but is not a primitive type)

It's fairly easy to build a query that will hit the limits of `long`, e.g - 
```
presto:sf100> explain select 1 from lineitem l1, lineitem l2, lineitem l3 , lineitem l4, lineitem l4;
                                                                                                                         Query Plan                                                                                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[PlanNodeId 13][_col0] => [expr:integer]                                                                                                                                                                                                           
         Estimates: {source: CostBasedSourceInfo, rows: 9,223,372,036,854,775,807 (8192PB), cpu: 388,922,817,995,845,400,000,000,000,000,000,000,000,000,000.00, memory: 0.00, network: 388,922,817,995,845,400,000,000,000,000,000,000,000,000,000.00}     
         _col0 := expr (1:16)  
```

In the above cross-join example using TPCH SF100 lineitem, the output row count would be `600037902^4
= 1.29632750431103743243e35` (this is incorrectly truncated to long max in the [TextRenderer](https://github.com/prestodb/presto/blob/03a2bd8bd31dcc51e0f61b389e32d63dde0f807a/presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java#L249))

Reply:
But row count is not a return type width. In the example you cite, my initial reaction would be to use BigInteger, and perhaps consider double if performance really requires it. However, returnTypeWidth fits inside a single byte. There's no reason I can see to use a double here. 

Unless maybe returnTypeWidth is not at all what I think it is? But if so, this method needs to be renamed to be clearer.

Reply:
Gotcha. Agreed about this return type.  

My point was broader about using doubles in the optimizer core stats classes. While VariableStatsEstimate#averageRowSize is unlikely to hit a long max, fields like PlanNodeStatsEstimate#totalSize which represents the estimate of the size of the data flow, can hit limits pretty easily (e.g the `8192PB` in the above query is also incorrect and actual size is much larger)

Reply:
Perhaps his comment is an answer to all your comments regarding double -> int/long across the code base: https://github.com/prestodb/presto/pull/23545#discussion_r1786094895